### PR TITLE
Added kafka producer connector

### PR DESCRIPTION
**Pull request by RKathees**

### CODE REVIEW

File: kafka-producer/org.wso2.carbon.connector/src/main/java/org/wso2/carbon/connector/KafkaProduce.java [hunk 1]
```diff
@@ -0,0 +1,101 @@
...
+ * Produce the messages to the kafka brokers
+ */
+public class KafkaProduce extends AbstractConnector {
+    public void connect(MessageContext messageContext) throws ConnectException {
+
+        SynapseLog log = getLog(messageContext);
+        log.auditLog("SEND : send message to  Broker lists");
+        //Get the producer with the configuration
+        Producer<String, String> producer = KafkaUtils
+                .getProducer(messageContext);
```
Comment (shafreenAnfar) [inline kafka-producer/org.wso2.carbon.connector/src/main/java/org/wso2/carbon/connector/KafkaProduce.java]: Why we have written this statement in two different lines ?

File: kafka-producer/org.wso2.carbon.connector/src/main/java/org/wso2/carbon/connector/KafkaUtils.java [hunk 2]
```diff
@@ -0,0 +1,144 @@
...
+        MessageFormatter messageFormatter;
+        messageFormatter = MessageProcessorSelector.getMessageFormatter(messageContext);
+        OutputStream out;
+        StringWriter stringWriter;
+        stringWriter = new StringWriter();
+        out = new WriterOutputStream(stringWriter, format.getCharSetEncoding());
+        try {
+            if (out != null) {
+                messageFormatter.writeTo(messageContext, format, out, true);
+                out.close();
```
Comment (Buddhima) [inline kafka-producer/org.wso2.carbon.connector/src/main/java/org/wso2/carbon/connector/KafkaUtils.java]: Generally, closing an outputstream need to do inside a finally block

File: kafka-producer/org.wso2.carbon.connector/src/main/java/org/wso2/carbon/connector/KafkaUtils.java [hunk 3]
```diff
@@ -0,0 +1,144 @@
...
+    }
+
+    /**
+     * The ProducerConfig class encapsulates the values required for establishing the connection with brokers such as the broker list, message
+     * partition class, serializer class for the message, and partition key,etc.
+     */
+    public static Producer<String, String> getProducer(MessageContext messageContext) {
+
+        Axis2MessageContext axis2mc = (Axis2MessageContext) messageContext;
+        axis2mc.getAxis2MessageContext();
```
Comment (Buddhima) [inline kafka-producer/org.wso2.carbon.connector/src/main/java/org/wso2/carbon/connector/KafkaUtils.java]: Any reason for the above line ?

File: kafka-producer/org.wso2.carbon.connector/src/main/java/org/wso2/carbon/connector/KafkaUtils.java [hunk 4]
```diff
@@ -0,0 +1,144 @@
...
+        producerConfigProperties.put(KafkaConnectConstants.BUFFER_MAX_MESSAGES,
+                bufferingMaxMessages);
+        producerConfigProperties.put(KafkaConnectConstants.NO_MESSAGE_BATCHED_PRODUCER,
+                batchNoMessages);
+        producerConfigProperties.put(KafkaConnectConstants.BUFFER_SIZE, sendBufferSize);
+        producerConfigProperties.put(KafkaConnectConstants.REQUEST_TIMEOUT, requestTimeout);
+        producerConfigProperties.put(KafkaConnectConstants.BUFFER_MAX_TIME, bufferingMaxTime);
+        producerConfigProperties.put(KafkaConnectConstants.ENQUEUE_TIMEOUT, enqueueTimeout);
+        producerConfigProperties.put(KafkaConnectConstants.CLIENT_ID, clientId);
+
```
Comment (Buddhima) [inline kafka-producer/org.wso2.carbon.connector/src/main/java/org/wso2/carbon/connector/KafkaUtils.java]: I think you can eliminate lots of duplications using a Map and ```properties.putAll(map);``` ?

File: kafka-producer/org.wso2.carbon.connector/src/main/java/org/wso2/carbon/connector/KafkaProduce.java (hunk 1 above)
Comment (katheesR) [inline kafka-producer/org.wso2.carbon.connector/src/main/java/org/wso2/carbon/connector/KafkaProduce.java]: Fixed in the PR - https://github.com/wso2/esb-connectors/pull/242

File: kafka-producer/org.wso2.carbon.connector/src/main/java/org/wso2/carbon/connector/KafkaUtils.java (hunk 2 above)
Comment (katheesR) [inline kafka-producer/org.wso2.carbon.connector/src/main/java/org/wso2/carbon/connector/KafkaUtils.java]: Fixed in the PR - https://github.com/wso2/esb-connectors/pull/242

File: kafka-producer/org.wso2.carbon.connector/src/main/java/org/wso2/carbon/connector/KafkaUtils.java (hunk 3 above)
Comment (katheesR) [inline kafka-producer/org.wso2.carbon.connector/src/main/java/org/wso2/carbon/connector/KafkaUtils.java]: Fixed in the PR - https://github.com/wso2/esb-connectors/pull/242

File: kafka-producer/org.wso2.carbon.connector/src/main/java/org/wso2/carbon/connector/KafkaUtils.java (hunk 4 above)
Comment (katheesR) [inline kafka-producer/org.wso2.carbon.connector/src/main/java/org/wso2/carbon/connector/KafkaUtils.java]: In the Map also,We have to put the all parameters.

File: kafka-producer/org.wso2.carbon.connector/src/main/java/org/wso2/carbon/connector/KafkaProduce.java [hunk 5]
```diff
@@ -0,0 +1,101 @@
...
+        String key = this.getKey(messageContext);
+        try {
+            String message = this.getMessage(messageContext);
+            send(producer, topic, key, message);
+        } catch (Exception e) {
+            log.error("Kafka producer connector : Error sending the message to broker lists ");
+            throw new ConnectException(e);
+        } finally {
+            //Close the producer pool connections to all kafka brokers.Also closes the zookeeper client connection if any
+            producer.close();
```
Comment (prabathabey) [inline kafka-producer/org.wso2.carbon.connector/src/main/java/org/wso2/carbon/connector/KafkaProduce.java]: Do we not have to handle null producers in this particular instance?

File: kafka-producer/org.wso2.carbon.connector/src/main/java/org/wso2/carbon/connector/KafkaUtils.java [hunk 6]
```diff
@@ -0,0 +1,144 @@
...
+     * partition class, serializer class for the message, and partition key,etc.
+     */
+    public static Producer<String, String> getProducer(MessageContext messageContext) {
+
+        Axis2MessageContext axis2mc = (Axis2MessageContext) messageContext;
+        axis2mc.getAxis2MessageContext();
+        String brokers = (String) axis2mc.getAxis2MessageContext()
+                .getOperationContext().getProperty("kafka.brokerList");
+        String serializationClass = (String) axis2mc.getAxis2MessageContext()
+                .getOperationContext().getProperty("kafka.serializationClass");
```
Comment (prabathabey) [inline kafka-producer/org.wso2.carbon.connector/src/main/java/org/wso2/carbon/connector/KafkaUtils.java]: Maybe it's good to introduce constants for all these property names?

File: kafka-producer/org.wso2.carbon.connector/src/main/java/org/wso2/carbon/connector/KafkaProduce.java (hunk 5 above)
Comment (katheesR) [inline kafka-producer/org.wso2.carbon.connector/src/main/java/org/wso2/carbon/connector/KafkaProduce.java]: Fixed in the PR - https://github.com/wso2/esb-connectors/pull/337

File: kafka-producer/org.wso2.carbon.connector/src/main/java/org/wso2/carbon/connector/KafkaUtils.java (hunk 6 above)
Comment (katheesR) [inline kafka-producer/org.wso2.carbon.connector/src/main/java/org/wso2/carbon/connector/KafkaUtils.java]: Yes.
All property names are moved to the constant variables.
Fixed in the PR - https://github.com/wso2/esb-connectors/pull/337

**Pull request merged**